6.62. Добавить Redux-хранилище. Ready for review

velandrey opened the pull request:
### Какую задачу решаем
https://github.com/ekbhast/destroy-this-bubbles/issues/62 Добавить Redux-хранилище: - данные хранятся и обрабатываются через Redux; - состояния приложения хранятся в сторе


### Скриншоты/видяшка
Видео относится к другому решёному реквесту, но поскольку визуально store ничего не меняет приложил видео как работает Профиль.
https://github.com/user-attachments/assets/bdb1e570-df09-4cd4-a924-3e35004bc3f8

### TBD (если есть)
1. Настроен store с редюсерами

2. Созданы слайсы:
userSlice - данные пользователя
notificationSlice - уведомления об ошибке или успехе
gameSlice - результаты игры (score, accuracy, totalTime)

3. Созданы кастомные хуки:
useAppDispatch / useAppSelector - хуки Redux
useNotification - управление уведомлениями
useGame - управление игровыми данными

4. Изменения в компонентах:
App.tsx - добавлен провайдер store и глобальный компонент уведомлений
LoginPage - в случае авторизации состояние отправляем в store. Включены уведомеления.
ProfilePage - данные профиля берём из store
GamePage - результаты игры сохраняем в store

5. Создан компонент:
NotificationContainer - система уведомлений



###  Инфо для команды:
// Получение данных
const user = useAppSelector(state => state.user);
const game = useAppSelector(state => state.game);
const loading = useAppSelector(state => state.loading);
const notifications = useAppSelector(state => state.notification);

// Отправка действий
const dispatch = useAppDispatch();
dispatch(setUser(userData));

// Уведомления
const { showSuccess, showError } = useNotification();
showSuccess('Успешно!');

// Игра
const { setGameResults, startGame } = useGame();
setGameResults({ score: 100, accuracy: 95, totalTime: 60 });

// Доступные данные из store
user - данные пользователя и авторизация
game - результаты игры (score, accuracy, time)
notification - система уведомлений

velandrey committed:
62 - Добавить Redux-хранилище
Добавлен Store, Provider

velandrey committed:
62 - Добавить Redux-хранилище
User slice, Loading slice
Store structure

velandrey committed:
62 - Добавить Redux-хранилище
Новые компоненты уведомлений об ошибках и загрузке вынесены в store

velandrey committed:
62 - Добавить Redux-хранилище
Состояние пользователя вынесено в store

velandrey committed:
62 - Добавить Redux-хранилище
Состояние результатов игры вынесено в store

velandrey committed:
62 - Добавить Redux-хранилище
Menu fix

velandrey committed:
62 - Добавить Redux-хранилище
Login fix

velandrey committed:
62 - Добавить Redux-хранилище
Fix Test in provider

eetanaM commented on packages/client/src/components/notificationContainer/notificationContainer.tsx:
```diff
@@ -0,0 +1,43 @@
+
+type NotificationContainerProps = {
+    autoHideDuration?: number;
+};
```
Классическое уже для нас замечание)
Типы по договоренности с префиксом `T`

eetanaM commented on packages/client/src/components/notificationContainer/notificationContainer.tsx:
Классная реализация показа уведомлений совместно с `useNotification,` возьму себе на вооружение, очень удобно в использовании!

eetanaM commented on packages/client/src/hooks/redux.ts:
```diff
@@ -0,0 +1,5 @@
+import { useDispatch, useSelector, TypedUseSelectorHook } from 'react-redux';
+
+export const useAppDispatch = () => useDispatch<AppDispatch>();
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```
Супер, что не забыл про пре-типизированные диспатч и селектор!)
P.S. С какой-то из версий react-redux (нашел, с 9.1.0 :) ), добавили методы `withTypes`, чтобы проще было объявить эти хуки:
```
export const useAppDispatch = useDispatch.withTypes<AppDispatch>()
export const useAppSelector = useSelector.withTypes<RootState>()
```

velandrey replied:
Да, действительно так проще. Но я не стал обновлять redux, подумал что раз 8я версия предустановлена, возможно это может иметь какое-то значение для других пакетов.

eetanaM commented on packages/client/src/store/slices/loadingSlice.ts:
Не замечание, а для обсуждения:
Быть может не выносить отдельным слайсом статус загрузки, а точечно в слайсах (где есть/будут ожидающие запросы) добавить свои статусы загрузки, чтобы не блокировать весь UI на всякое ожидание, а блокировать частично только те компоненты, которые ожидают данных, связанных с отдельным слайсом?

eetanaM commented on packages/client/src/store/slices/gameSlice.ts:
Опять же для обсуждения:
Думается, что есть смысл добавить в слайс игры, помимо результатов и состояния игры, объект с настройками игры, с начальным состоянием из конфига `gameSettings`, который @ekbhast подготовил, и которые пользователь сможет менять на странице с игрой. А для обсуждения потому, что надо до конца определиться с тем, будут ли у нас отдельные режимы, если да, то какие и/или насколько можно будет пользователю кастомизировать процесс игры в целом, возможно нам будет достаточно одного радио-инпута на самой странице с игрой с выбором между двумя режимами, например, и тогда можно не усложнять слайс

eetanaM commented on packages/client/src/store/slices/notificationSlice.ts:
```diff
@@ -0,0 +1,46 @@
+
+type NotificationState = {
+    notifications: Notification[];
+};
```
кастомные типы с `T` - префиксом :)

eetanaM commented on packages/client/src/store/slices/gameSlice.ts:
```diff
@@ -0,0 +1,48 @@
+type GameState = {
+    results: GameResults | null;
+    isGameActive: boolean;
+};
```
кастомные типы с `T` - префиксом :)

eetanaM reviewed (commented):
В целом всё супер, redux настроен хорошо! 👍
Есть несколько вопросов для обсуждения, все в комментах к файлам, и пара микро-недочетов по неймингу типов, в остальном отличная работа!

UPD: Можно было бы ещё все асинхронные функции обернуть в `createAsyncThunk`, чтобы вся логика взаимодействия с хранилищем была централизована и описывалась прямо в слайсах в `extraReducers`, а не была размазана по компонентам, в компонентах же были бы простые вызовы нужных экшенов

velandrey committed:
62 - Добавить Redux-хранилище
Fix Types Name

velandrey committed:
62 - Добавить Redux-хранилище
merge develop

velandrey committed:
62 - Добавить Redux-хранилище
merge develop

velandrey committed:
62 - Добавить Redux-хранилище
merge develop

velandrey committed:
62 - Добавить Redux-хранилище
Добавлен createAsyncThunk
Удалён лоадер
Оптимизировано состояние для уведомлений

velandrey commented:
Пофиксил все замечания, влил актуальную ветку develop, добавил createAsyncThunk - сейчас с ним код кажется стал более логичным и читабельным

velandrey committed:
Merge branch 'develop' into 62-redux

# Conflicts:
#	packages/client/src/pages/game/game.tsx

velandrey committed:
62 - Добавить Redux-хранилище
Слияние ветки develop

velandrey committed:
62 - Добавить Redux-хранилище
Слияние ветки develop

eetanaM reviewed (approved):
Хорошая работа!
Единственное, уже когда наверное будем причесывать проект, нужно будет убрать слайс загрузки вообще (либо, если все же решим оставить общий статус загрузки, то надо будет вернуть его в конфигурацию store) и кастомные хуки (как useProfile, например), которые уже тоже не понадобятся, с учетом того, что всё делаем на asyncThunk'ах, потому что сейчас они уже кажутся просто лишним слоем.

kyzinatra commented on packages/client/src/hooks/useGame.ts:
```diff
@@ -0,0 +1,35 @@
+        settings,
+        setGameResults: (results: TGameResults) =>
+            dispatch(setGameResults(results)),
+        startGame: () => dispatch(startGame()),
```
события лучше вынести в отдельный actions. Можно использовать createAction. Там же можно взять данные из redux. Тогда не нужен будет этот хук

kyzinatra commented on packages/client/src/hooks/useNotification.ts:
```diff
@@ -0,0 +1,44 @@
+
+export const useNotification = () => {
+    const dispatch = useAppDispatch();
+    const { notifications } = useAppSelector((state) => state.notification);
```
Лучше более специфично const notifications = useAppSelector((state) => state.notification.notifications );

kyzinatra commented on packages/client/src/hooks/useNotification.ts:
```diff
@@ -0,0 +1,44 @@
+        showSuccess: (message: string, autoHideDuration?: number) =>
+            showNotification(message, 'success', autoHideDuration),
+        showWarning: (message: string, autoHideDuration?: number) =>
+            showNotification(message, 'warning', autoHideDuration),
```
по аналогии стоит вынести в `createAction` или вовсе прокидывать их из createSlice

kyzinatra reviewed (approved):
Отличная работа! Хорошая интеграция с redux! Я бы вынес события отдельно, чтобы не создавать их в отдельном хуке и не заставлять компоненты обновляться, чтобы актуализировать состояние, но в целом работа отличная

velandrey commented:
Влил в develop store. Дополнения лучше отдельным фикс-реквестом завезу.
